packages/realm/src/schema/types.ts

```diff
@@ -71,30 +101,221 @@ export type CanonicalObjectSchemaProperty = {
+   * case of a collection, to be assigned to its elements. (Realm object types in lists
+   * and sets cannot be optional.)
+   *
+   * Default value: See the requirements in the documentation for this type.
+   */
+  optional?: boolean;
+  /**
+   * Whether the property should be indexed. (Only supported if `type` is `"string"`,
+   * `"int"`, or `"bool"`).
+   *
+   * Default value: `false` if the property is not a primary key, otherwise `true`.
+   */
+  indexed?: boolean;
+  /**
+   * The name to be persisted in the Realm file if it differs from the already-defined
+   * JavaScript/TypeScript (JS/TS) property name. This is useful for allowing different
+   * naming conventions than what is persisted in the Realm file. Reading and writing
+   * properties must be done using the JS/TS name, but queries can use either the JS/TS
+   * name or the persisted name.
+   */
+  mapTo?: string;
+  /**
+   * The default value that the property will be set to when created.
+   */
+  default?: unknown;
+};
+
+/**
+ * Keys used in the property schema that are common among all variations of {@link PropertySchemaStrict}.
+ */
+type PropertySchemaCommon = {
+  indexed?: boolean;
+  mapTo?: string;
+  default?: unknown;
+};
+
+/**
+ * The strict schema for specifying the type of a specific Realm object property.
+ *
+ * Unlike the less strict {@link PropertySchema}, this type precisely defines the type
+ * requirements and their allowed combinations; however, TypeScript error messages tend
+ * to be more misleading. {@link PropertySchema} is recommended for that reason, but the
+ * strict type is provided as guidance.
+ *
+ * @see {@link PropertySchema} for a textual explanation of the requirements defined here,
+ *   as well as documentation for each property.
+ */
+export type PropertySchemaStrict = PropertySchemaCommon &
+  (
+    | {
+        type: Exclude<PrimitivePropertyTypeName, "mixed">;
+        optional?: boolean;
+        // TODO: Do we want TS errors for PropertySchemaStrict when `objectType` and
+        //       `property` exist on the schema as `undefined`? (The parser allows this)
+        // objectType?: undefined;
+        // property?: undefined;
+      }
+    | {
+        type: "mixed";
+        optional?: true;
+        // objectType?: undefined;
+        // property?: undefined;
+      }
+    | {
+        type: CollectionPropertyTypeName;
+        objectType: Exclude<PrimitivePropertyTypeName, "mixed">;
+        optional?: boolean;
+        // property?: undefined;
+      }
+    | {
+        type: CollectionPropertyTypeName;
+        objectType: "mixed";
+        optional?: true;
+        // property?: undefined;
+      }
+    | {
+        type: "list" | "set";
+        objectType: UserTypeName;
+        optional?: false;
+        // property?: undefined;
+      }
+    | {
+        type: "dictionary";
+        objectType: UserTypeName;
+        optional?: true;
+        // property?: undefined;
```

Comment:
Just curious why these are commented out?

Reply:
It had to do with the `TODO` comment (should've made that clearer :p) If we don't have those fields, then TS will generate errors if trying to set them (even if the value is `undefined` which is allowed at runtime). If we add them (but only allow `undefined`) then the rules align with our parser.